diff --git a/node_modules/@changesets/assemble-release-plan/dist/assemble-release-plan.cjs.dev.js b/node_modules/@changesets/assemble-release-plan/dist/assemble-release-plan.cjs.dev.js
index de55d40..afd5e5d 100644
--- a/node_modules/@changesets/assemble-release-plan/dist/assemble-release-plan.cjs.dev.js
+++ b/node_modules/@changesets/assemble-release-plan/dist/assemble-release-plan.cjs.dev.js
@@ -325,7 +325,7 @@ function _defineProperty(obj, key, value) { if (key in obj) { Object.definePrope
 
 function getPreVersion(version) {
   let parsed = semver.parse(version);
-  let preVersion = parsed.prerelease[1] === undefined ? -1 : parsed.prerelease[1];
+  let preVersion = -1;
 
   if (typeof preVersion !== "number") {
     throw new errors.InternalError("preVersion is not a number");
