diff --git a/node_modules/jest-cli/node_modules/jest-snapshot/build/InlineSnapshots.js b/node_modules/jest-cli/node_modules/jest-snapshot/build/InlineSnapshots.js
index 5f9c3d5..1ec2ce5 100644
--- a/node_modules/jest-cli/node_modules/jest-snapshot/build/InlineSnapshots.js
+++ b/node_modules/jest-cli/node_modules/jest-snapshot/build/InlineSnapshots.js
@@ -219,7 +219,8 @@ const saveSnapshotsForFile = (snapshots, sourceFilePath, prettier) => {
     filename: sourceFilePath,
     plugins,
     presets,
-    root: path.dirname(sourceFilePath)
+    // Jest 27-28 currently can't locate our `babel.config.js` correctly, see: https://github.com/facebook/jest/issues/11741
+    // root: path.dirname(sourceFilePath)
   });
 
   if (!ast) {
